workflow/common/validate.go

```diff
@@ -38,93 +52,60 @@ func ValidateWorkflow(wf *wfv1.Workflow) error {
-	// Validate Template Name
-	if err := validateFieldName(tmpl.Name, "workflow.spec.templates.%s has invalid name: %s"); err != nil {
+	err = ctx.validateTemplate(entryTmpl, ctx.wf.Spec.Arguments)
+	if err != nil {
 		return err
 	}
-	// Validate globalParameter names
-	for _, globalParam := range wfGlobalParameters {
-		if err := validateFieldName(globalParam.Name, "workflow.spec.arguments.parameters.%s has invalid name: %s"); err != nil {
-			return err
+	if ctx.wf.Spec.OnExit != "" {
+		exitTmpl := ctx.wf.GetTemplate(ctx.wf.Spec.OnExit)
+		if exitTmpl == nil {
+			return errors.Errorf(errors.CodeBadRequest, "spec.onExit template '%s' undefined", ctx.wf.Spec.OnExit)
 		}
-	}
-	// Validate Argument Names
-	for _, argParam := range args.Parameters {
-		if err := validateFieldName(argParam.Name, "arguments.parameters.%s has invalid name: %s"); err != nil {
-			return err
-		}
-	}
-	for _, argArt := range args.Artifacts {
-		if err := validateFieldName(argArt.Name, "arguments.artifacts.%s has invalid name: %s"); err != nil {
-			return err
-		}
-	}
-	// Validate Input Names
-	for _, inParam := range tmpl.Inputs.Parameters {
-		if err := validateFieldName(inParam.Name, "inputs.parameters.%s has invalid name: %s"); err != nil {
-			return err
-		}
-	}
-	for _, inArt := range tmpl.Inputs.Artifacts {
-		if err := validateFieldName(inArt.Name, "inputs.artifacts.%s has invalid name: %s"); err != nil {
-			return err
-		}
-	}
-	// Validate Output Names
-	for _, outParam := range tmpl.Outputs.Parameters {
-		if err := validateFieldName(outParam.Name, "outputs.parameters.%s has invalid name: %s"); err != nil {
-			return err
-		}
-	}
-	for _, outArt := range tmpl.Outputs.Artifacts {
-		if err := validateFieldName(outArt.Name, "outputs.artifacts.%s has invalid name: %s"); err != nil {
+		// now when validating onExit, {{workflow.status}} is now available as a global
+		ctx.globalParams["workflow.status"] = "placeholder-workflow-status"
+		err = ctx.validateTemplate(exitTmpl, ctx.wf.Spec.Arguments)
+		if err != nil {
 			return err
 		}
 	}
-	// Validate step names
-	for _, stepList := range tmpl.Steps {
-		for _, step := range stepList {
-			if err := validateFieldName(step.Name, "steps.%s has invalid name: %s"); err != nil {
-				return err
-			}
-		}
-	}
 	return nil
 }
 
-func (ctx *wfValidationCtx) validateTemplate(tmpl *wfv1.Template, args wfv1.Arguments, wfGlobalParameters []wfv1.Parameter) error {
-	err := validateTemplateFieldNames(tmpl, args, wfGlobalParameters)
-	if err != nil {
-		return err
-	}
+func (ctx *wfValidationCtx) validateTemplate(tmpl *wfv1.Template, args wfv1.Arguments) error {
 	_, ok := ctx.results[tmpl.Name]
 	if ok {
 		// we already processed this template
 		return nil
 	}
 	ctx.results[tmpl.Name] = true
-	_, err = ProcessArgs(tmpl, args, wfGlobalParameters, true)
+	_, err := ProcessArgs(tmpl, args, ctx.globalParams, true)
```

Comment:
My only question is if there are any issues with doing ProcessArgs before doing all field name validates (input/outupt/steps/etc at this point we have already validated args and template names) since ProcessArgs uses the names and if we will end up in any situation where the error reported is from ProcessArgs which is obtuse when the issue is name field.

Reply:
Fair point. I am going ensure validateArguments is called prior to the call to validateTemplate()